slacksocketmode/socket_mode_managed_conn.go

```diff
@@ -0,0 +1,441 @@
+//
+// How it works:
+//
+// 1. The handler stops if there is any "signal" sent from within this Client
+// 2. The handler stops if Slack stopped sending ping in a timely manner
+// 3. Sends outgoing messages that are received from the Client's socketModeResponses channel
+// 4. Handles incoming raw events from the webSocketMessages channel.
+func (smc *Client) runMessageHandler(webSocketMessages chan json.RawMessage) {
+	ticker := time.NewTicker(smc.pingInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		// 1. catch "stop" signal on channel close
+		case intentional := <-smc.killChannel:
+			_ = smc.killConnection(intentional, errorsx.String("signaled"))
+			return
+		// 2. detect when the connection is dead.
+		case <-smc.pingDeadman.C:
+			_ = smc.killConnection(false, slack.ErrRTMDeadman)
+			return
+		// 3. listen for messages that need to be sent
+		case res := <-smc.socketModeResponses:
+			if err := smc.unsafeWriteSocketModeResponse(res); err != nil {
+				smc.IncomingEvents <- smc.internalEvent(EventTypeErrorWriteFailed, &ErrorWriteFailed{
+					Cause:    err,
+					Response: res,
+				})
+			}
+			// listen for incoming messages that need to be parsed
+		case wsMsg := <-webSocketMessages:
+			_ = smc.handleWebSocketMessage(wsMsg)
+		}
+	}
+}
+
+// runMessageReceiver monitors the Socket Mode opened WebSocket connection for any incoming
+// messages. It pushes the raw events into the channel.
+//
+// This will stop executing once the RTM's when a fatal error is detected, or
+// a disconnect occurs.
+func (smc *Client) runMessageReceiver(sink chan json.RawMessage) {
+	for {
+		if err := smc.receiveMessagesInto(sink); err != nil {
+			select {
+			case smc.killChannel <- false:
+			case <-smc.disconnected:
+			}
+			return
+		}
+	}
+}
+
+type Response struct {
+	EnvelopeID string `json:"envelope_id"`
+}
+
+// unsafeWriteSocketModeResponse sends a WebSocket message back to Slack.
+// WARNING: Call to this function must be serialized!
+//
+// Here's why - Gorilla WebSocket's Writes functions are not concurrency-safe.
+// That is, we must serialize all the writes to it with e.g. a goroutine or mutex.
+// We intentionally chose to use goroutine, which makes it harder to propagate write errors to the caller,
+// but is more computationally efficient.
+//
+// See the below for more information on this topic:
+// https://stackoverflow.com/questions/43225340/how-to-ensure-concurrency-in-golang-gorilla-websocket-package
+func (smc *Client) unsafeWriteSocketModeResponse(msg *Response) error {
+	// set a write deadline on the connection
+	if err := smc.conn.SetWriteDeadline(time.Now().Add(10 * time.Second)); err != nil {
+		return err
+	}
+
+	// Remove write deadline regardless of WriteJSON succeeds or not
+	defer smc.conn.SetWriteDeadline(time.Time{})
+
+	if err := smc.conn.WriteJSON(msg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (smc *Client) internalEvent(tpe string, data interface{}) ClientEvent {
```

Comment:
What is the difference between `internalEvent` and `externalEvent`?

Reply:
`tpe` is another form of `type` you occasionally use when you want a variable for some kind of type but you need to avoid conflict with the reserved keyword `type`

Reply:
Re `tpe`, I'm fine to hear your opinion for alternative name but anyway, you can see it used widely https://github.com/search?l=Go&q=%22tpe%22&type=Code

Reply:
Thanks for pointing it out.

`internalEvent` vs `externalEvent` was some artifact left while I was scratching my head structuring code. It's not necessary anymore and removed in 11f20f31b34a0c2b7a80a21c794f4ea96865bc4a!

Reply:
> tpe is another form of type you occasionally use when you want a variable for some kind of type but you need to avoid conflict with the reserved keyword type

That was very educational. Thank you :)